Complete separation of Data and Codespace

comcon1 opened the pull request:
This refactoring introduces several new features:
- simulation dict is packed into `System` class which works as a dict but extends it
- molecules are packed into specific class `Lipid` and `NonLipid` both inherit from `Molecule`
- `Molecule` can register mapping file for it and then return mapping dictionary by request
```
ss = initialize_databank()
s: System = ss.loc(200)  # this is System object
sys_dict: dict = s.readme  # this is the dictionary
lip: Lipid = s.content['POPC']  # this is object of Lipid class
for k, v in lip.mapping_dictionary.items():  # this is how mapping is loaded
   print(k)
   print(v)
```
- `lipids_dict` is deprecated, we use `lipids_set` instead. It's a `LipidSet(MoleculeSet)` object which can check strings and Molecules
```
form DatabankLib import lipids_set
assert "POPC" in lipids_set
assert lip in lipids_set  # you can also check Lipid object by the same command
```
It is constructed dynamically at the stage of module import from `Data/Molecules`.

batukav commented:
I’ll reply with my comments in the next 1-2 days.

> On 19. May 2025, at 13:51, Alexey Nesterenko ***@***.***> wrote:
> 
> 
> @comcon1 <https://github.com/comcon1> requested your review on: #227 <https://github.com/NMRLipids/Databank/pull/227> Complete separation of Data and Codespace.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/NMRLipids/Databank/pull/227#event-17712558795>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB47SNPAW7ITJCQENXA5TD327GZSJAVCNFSM6AAAAAB5NMNKRWVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJXG4YTENJVHA3TSNI>.
> You are receiving this because your review was requested.
>

MagnusSletten reviewed (commented):
Not close to being done reviewing the changes yet, but I really like what i'm seeing. Using inheritance in code can sometimes get unwieldy over time, but here with how you handle Molecules, Lipids and Non-lipids I thought it was great and it's a completely natural way to do it. 
I really like the flexible solution of default data locations, but with optional enviromental variable which overrides it. It makes testing a lot cleaner. 
Haven’t seen any errors of consequence yet. I will continue tomorrow.

MagnusSletten commented on Scripts/DatabankLib/settings/molecules.py:
```diff
@@ -6,431 +6,297 @@
+        """
+        if self._test_item_type(item):
+            self._items.add(item)
+            self._names.add(item.name.toupper())
```
You use the method toupper() twice, but I think maybe you just mean to use upper() here since toupper() is not a standard string method.

comcon1 replied:
commit 50b0c138 solves it

batukav commented on Scripts/BuildDatabank/searchDATABANK.py:
```diff
@@ -8,141 +8,131 @@
             try:
                 if self.readme['ION_CONCENTRATIONS'][key] != 0:
-                    expIons.append(key)
+                    exp_ions.append(key)
```
should we try to catch other exceptions?

comcon1 replied:
mm.. No probably. I widened a bit allowed Exception list in the next block in b249796d

batukav commented on Scripts/BuildDatabank/searchDATABANK.py:
```diff
@@ -152,82 +142,82 @@ def loadSimulations() -> List[Simulation]:
+    for subdir in tqdm(rm_idx, desc='Experiment'):
+        exp_readme_fp = os.path.join(subdir, 'README.yaml')
+        with open(exp_readme_fp) as yaml_file_exp:
+            exp_readme = yaml.load(yaml_file_exp, Loader=yaml.FullLoader)
```
I suggest putting a try/except block here

comcon1 replied:
I added FileNotFoundError, PermissionError processing as they are not super severe for whole script work. Fixed in b249796d

batukav commented on Scripts/DatabankLib/core.py:
```diff
@@ -59,6 +59,8 @@ def __init__(self, data=None):
             elif k in molecules_set:
                 mol = NonLipid(k)
+            else:
+                raise ValueError(f"Molecule {k} is not in the set of lipids or molecules.")
```
I'd suggest putting this into a try/except block.

batukav commented on Scripts/DatabankLib/core.py:
```diff
@@ -111,11 +113,18 @@ def __load_systems__(self):
+                    except FileNotFoundError as e:
+                        print(f"Problem loading mapping file for the system: {e}")
+                        print(f"System ID: {ydict['ID']}")
+                        print(f"System path: {subdir}")
```
add 

```
except Exception as e:
    print(f"An unexpected error occurred: {e}")
    print(f"System ID: {ydict.get('ID', 'Unknown')}")
    print(f"System path: {subdir}")
```

for unforeseen exceptions.

MagnusSletten commented:
One thing to note is that it's more important with this setup to run tests directly from the tests folder because otherwise Databanklib won't get the mocked paths correctly. If started from root of repo then it seems like  Databanklib is imported prior to tests and uses standard data paths which causes large files to be downloaded.

Maybe we make the documentation a bit more clear on this or change the behaviour. 

From root:

```
Scripts/tests/test_AddData.py::TestAddData::test_add_data_h DBG: Mocking Data path:  /Databank/Data
DBG: Mocking Simulations path:  /Databank/Data/Simulations
STDOUT usage: AddData.py Script [-h] [-f FILE] [-d] [-n] [-w WORK_DIR] [-o OUTPUT_DIR]
```

from tests (works fine):

```
test_AddData.py::TestAddData::test_add_data_h DBG: Mocking Data path:  /Databank/Scripts/tests/Data
DBG: Mocking Simulations path:  /Databank/Scripts/tests/Data/Simulations.1
STDOUT usage: AddData.py Script [-h] [-f FILE] [-d] [-n] [-w WORK_DIR] [-o OUTPUT_DIR]
```

ohsOllila commented:
>molecules are packed into specific class Lipid and NonLipid both inherit from Molecule

Would it make sense to have already also proteins class?

ohsOllila commented:
I will present NMRlipids databank with hands on session next week in here: https://www.cecam.org/workshop-details/machine-learning-in-physical-sciences-theory-and-applications-1449

I was planning to use Colab notebooks from databank-template repo there. When planning to merge this, we should make sure that they work at least next week.

comcon1 commented on Scripts/DatabankLib/quality.py:
```diff
@@ -16,39 +17,32 @@
-    def __init__(self, readme, op_data, ff_data, idx_path):
-        self.readme = readme.copy()
+    def __init__(self, system: System, op_data, ff_data, idx_path):
+        self.system: System = system
```
I started to use type hints almost everywhere as I found it important for IDEs and Jupyter notebooks. If you have type hints, code-autocompleter will give you proper hints and autocomplete attributes. Now, for the current code the required python version is 3.9 as it is a threshold for most important typing improvements as I found. Some code parts are still not organized in a proper way but at least core API must be covered. So if you see something in core API that is not covered, you can report. 

This `quality` file is actually still purely revised. It should be improved more significantly and probably added to API because it looks natural to have some quality evaluation stuff in API.

comcon1 replied:
I created a new subissue #233 describing typehints in the codestyle guide.

comcon1 commented:
> If I now use the current version of AddData.py, is it complicate to update to the new one? In principle it might be better to wait the update, but I would like to progress now as I have some time for it.

You can just fork from my branch, which is pull-requested here, and work starting from this point. At least, the code is separated in this branch.

> Would it make sense to have already also proteins class?

I still don't know if one code can cover both protein and lipids. But you could at least inherit the architecture.

> I was planning to use Colab notebooks from databank-template repo there. When planning to merge this, we should make sure that they work at least next week.

OK, then we will probably not be able to completely divide the repository next week. But Colab notebooks will keep working. Will make sure.

comcon1 commented:
> One thing to note is that it's more important with this setup to run tests directly from the tests folder because otherwise Databanklib won't get the mocked paths correctly. If started from root of repo then it seems like Databanklib is imported prior to tests and uses standard data paths which causes large files to be downloaded.
> 
> Maybe we make the documentation a bit more clear on this or change the behaviour.

Oh, again. That I didn't know, thanx. It's a very fragile thing. I'm aware that if to import in the header of a test-module, it always happens, so I wrote it explicitly in every header of every test-module. If something is done wrong, it starts using a standard folder immediately and it's annoying.. I think now that I have invented some solution around it: let's add to data directory `.nofortest` and put small readme inside, which writes that the content of Data folder should not be used in unit-tests.

In the autostarting functions of test-modules we test if directory `.nofortest` does not exist in `NMLDB_DATA_PATH`. 

Using `Data` in tests must be completely forbidden, as it will becomes a new repository, and tests cannot depend on a different repository.

MagnusSletten commented:
> Oh, again. That I didn't know, thanx. It's a very fragile thing. I'm aware that if to import in the header of a test-module, it always happens, so I wrote it explicitly in every header of every test-module. If something is done wrong, it starts using a standard folder immediately and it's annoying.. I think now that I have invented some solution around it: let's add to data directory `.nofortest` and put small readme inside, which writes that the content of Data folder should not be used in unit-tests.
> 
> In the autostarting functions of test-modules we test if directory `.nofortest` does not exist in `NMLDB_DATA_PATH`.
> 
> Using `Data` in tests must be completely forbidden, as it will becomes a new repository, and tests cannot depend on a different repository.

I wonder if maybe there's a simple solution by adding a direct reset path method to the init.py file within Databanklib. Then you can use this method in the header of tests to be absolutely sure they are set properly because this solution does not depend on order of operations.

comcon1 commented:
You mean direct reset by re-reading enviroment? Is it possible to reset globally defined static variables? Will it be surely properly affected? Because we have, for example dictionaries which are constructed from Data/Molecules. They should be also reset. The stable solution is probably to hang everything on classes (including global configs), use singletons and describe dependencies to ensure order. It's a bit hard now I feel.

On Thursday, May 22nd, 2025 at 9:11 AM, MagnusSletten ***@***.***> wrote:

> MagnusSletten left a comment [(NMRLipids/Databank#227)](https://github.com/NMRLipids/Databank/pull/227#issuecomment-2900158987)
>
>> Oh, again. That I didn't know, thanx. It's a very fragile thing. I'm aware that if to import in the header of a test-module, it always happens, so I wrote it explicitly in every header of every test-module. If something is done wrong, it starts using a standard folder immediately and it's annoying.. I think now that I have invented some solution around it: let's add to data directory .nofortest and put small readme inside, which writes that the content of Data folder should not be used in unit-tests.
>>
>> In the autostarting functions of test-modules we test if directory .nofortest does not exist in NMLDB_DATA_PATH.
>>
>> Using Data in tests must be completely forbidden, as it will becomes a new repository, and tests cannot depend on a different repository.
>
> I wonder if maybe there's a simple solution by adding a direct reset path method to the init.py file within Databanklib. Then you can use this method in the header of tests to be absolutely sure they are set properly because this solution does not depend on order of operations.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/NMRLipids/Databank/pull/227#issuecomment-2900158987), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AGBXJOGBLT25VLIRGKS5FSL27V2DJAVCNFSM6AAAAAB5NMNKRWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMBQGE2TQOJYG4).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

MagnusSletten commented:
> You mean direct reset by re-reading enviroment? Is it possible to reset globally defined static variables? Will it be surely properly affected? Because we have, for example dictionaries which are constructed from Data/Molecules. They should be also reset. The stable solution is probably to hang everything on classes (including global configs), use singletons and describe dependencies to ensure order. It's a bit hard now I feel.

Yeah, there are several ways to go about this and it's good think carefully through this. I agree the singleton approach is probably best as it is a pretty standard way to go about it. I haven't thought enough about proper use here for me to strongly recommend a specific approach but I did try what I outlined:

Simple setup with added method to the init file:

```
def reset_data_paths():
    global NMLDB_DATA_PATH, NMLDB_SIMU_PATH 

    NMLDB_DATA_PATH =  os.environ.get(
    "NMLDB_DATA_PATH",
    os.path.join(NMLDB_ROOT_PATH, 'Data'))
    
    NMLDB_SIMU_PATH = os.environ.get(
    "NMLDB_SIMU_PATH",
    os.path.join(NMLDB_DATA_PATH, 'Simulations')
)
```
The global line there just makes sure the variables aren't just changed within the scope of the method.

Then in tests for the header method I just add the method:

```
@pytest.fixture(autouse=True, scope="module")
def header_module_scope():
    os.environ["NMLDB_DATA_PATH"] = os.path.join(os.path.dirname(__file__), "Data")
    os.environ["NMLDB_SIMU_PATH"] = os.path.join(os.path.dirname(__file__), "Data", "Simulations.1")
    import DatabankLib
    DatabankLib.reset_data_path()
    print("DBG: Mocking Data path: ", DatabankLib.NMLDB_DATA_PATH)
    print("DBG: Mocking Simulations path: ", DatabankLib.NMLDB_SIMU_PATH)
    yield
    print("DBG: Mocking completed")

```
It did work and I didn't see behaviour that was unintended. This would also work with dictionaries related to Molecules if the method creating the dict  (load_from_data) is called in the same script after changing Mol paths. 

Alternative can be reloading the entire module with: [importlib](https://docs.python.org/3/library/importlib.html) and using the importlib.reload(module) method. This is something I have not tried.

comcon1 commented:
I will isolate this part of the discussion into the separated issue.

comcon1 commented on Scripts/DatabankLib/settings/molecules.py:
```diff
@@ -6,431 +6,297 @@
+    """
+    Class for non-bilayer molecules: solvent, ions, etc.
+    """
+    _nonlipids_dir: str = os.path.join(NMLDB_MOL_PATH, "env")
```
will rename to "membrane" and "solution". Because in membrane there can be non-lipid molecules as well. In solution, it can be ions, solvent molecules and non-ionic soluble compounds.

comcon1 replied:
commit 50b0c138 solves it

comcon1 commented on Scripts/DatabankLib/settings/molecules.py:
```diff
@@ -6,431 +6,297 @@
+    @staticmethod
+    def load_from_data():
+        """
+        Loads lipid data from the designated directory and returns a set of lipids.
```
commit 50b0c138 solves it

comcon1 commented:
> I wonder if maybe there's a simple solution by adding a direct reset path method to the init.py file within Databanklib. Then you can use this method in the header of tests to be absolutely sure they are set properly because this solution does not depend on order of operations.

I significantly improved testing part. I added subissue #234 for description about how we write unit-tests.

Now there are three blocks of tests:
```
pytest --cmdopt sim2
pytest --cmdopt sim1
pytest --cmdopt nodata
```
they run only specific test files and initialize environmental variables according to cmdopt. So simulation.2 tests will be skipped in sim1 by a hook in `conftest.py`. This is fixed in commit 81c39501.

I also implemented `.notest` defence in addition in the same commit, but I don't see that something now activates it.

comcon1 commented:
> Is it intended behaviour then that pytest cannot be invoked from root directory?

I cannot reproduce the bug. I run it from root and everything works fine. Path is mocked properly. Please describe your setup for reproducibility. I try to invoke it from root path, and it works.

MagnusSletten commented:
> I cannot reproduce the bug. I run it from root and everything works fine. Path is mocked properly. Please describe your setup for reproducibility. I try to invoke it from root path, and it works.

Here is setup details and the runs of tests: 

[Setup details](https://gist.github.com/MagnusSletten/b41bff22aa72f941a382cbc74c6a473e
)


I checkout the branch within docker and run the dev version of pip install you described. Then from root I test both plain "pytest -vs" and more explicit: "pytest --cmdopt sim1"

MagnusSletten commented:
I commented on the test setup in the gist link and I also see you fixed the other very minor errors I pointed out. From my perspective this PR is solid but I see Butahan had some comments about improvements too.

ohsOllila commented:
When running 

pip install -e .

I get the error below. Is this a bug or have I messed up my environments?


Installing collected packages: DatabankLib
  Running setup.py develop for DatabankLib
    error: subprocess-exited-with-error
    
    × python setup.py develop did not run successfully.
    │ exit code: 1
    ╰─> [32 lines of output]
        running develop
        /usr/lib/python3/dist-packages/setuptools/command/easy_install.py:158: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
          warnings.warn(
        WARNING: The user site-packages directory is disabled.
        /usr/lib/python3/dist-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
          warnings.warn(
        error: can't create or remove files in install directory
        
        The following error occurred while trying to add or remove files in the
        installation directory:
        
            [Errno 13] Permission denied: '/usr/local/lib/python3.10/dist-packages/test-easy-install-39754.write-test'
        
        The installation directory you specified (via --install-dir, --prefix, or
        the distutils default setting) was:
        
            /usr/local/lib/python3.10/dist-packages/
        
        Perhaps your account does not have write access to this directory?  If the
        installation directory is a system-owned directory, you may need to sign in
        as the administrator or "root" account.  If you do not have administrative
        access to this machine, you may wish to choose a different installation
        directory, preferably one that is listed in your PYTHONPATH environment
        variable.
        
        For information on other options, you may wish to consult the
        documentation at:
        
          https://setuptools.pypa.io/en/latest/deprecated/easy_install.html
        
        Please make the appropriate changes for your system and try again.
        
        [end of output]
    
    note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× python setup.py develop did not run successfully.
│ exit code: 1
╰─> [32 lines of output]
    running develop
    /usr/lib/python3/dist-packages/setuptools/command/easy_install.py:158: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
      warnings.warn(
    WARNING: The user site-packages directory is disabled.
    /usr/lib/python3/dist-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
      warnings.warn(
    error: can't create or remove files in install directory
    
    The following error occurred while trying to add or remove files in the
    installation directory:
    
        [Errno 13] Permission denied: '/usr/local/lib/python3.10/dist-packages/test-easy-install-39754.write-test'
    
    The installation directory you specified (via --install-dir, --prefix, or
    the distutils default setting) was:
    
        /usr/local/lib/python3.10/dist-packages/
    
    Perhaps your account does not have write access to this directory?  If the
    installation directory is a system-owned directory, you may need to sign in
    as the administrator or "root" account.  If you do not have administrative
    access to this machine, you may wish to choose a different installation
    directory, preferably one that is listed in your PYTHONPATH environment
    variable.
    
    For information on other options, you may wish to consult the
    documentation at:
    
      https://setuptools.pypa.io/en/latest/deprecated/easy_install.html
    
    Please make the appropriate changes for your system and try again.
    
    [end of output

MagnusSletten commented:
>  [Errno 13] Permission denied: '/usr/local/lib/python3.10/dist-packages/test-easy-install-39754.write-test'

It might be some change in your local environment causing permission issues. If you do a quick test with something like "pip install numpy", does it work well? Just a quick test to see if pip can install anything without error in your enviroment.

comcon1 commented:
> [Errno 13] Permission denied: '/usr/local/lib/python3.10/dist-packages/test-easy-install-39754.write-test'

Looks like your user doesn't have permissions to install it to your dist-packages. If you have conda environment, then it definitely doesn't use it, because it uses folder in `/usr/local/lib`. To check if you actually use the environment, run the following command before you are doing `pip install`:
```
~/repo/Databank$ which pip
/home/comcon1/.conda/envs/databankdev2/bin/pip
```
You see, in my case it refers to pip in `~/.conda/envs/...` so it is definitely conda's pip. 

Just in case, I did right now:
```
conda create -n databankdev2 -c conda-forge python=3.10
conda activate databankdev2
pip install -e .
```
And everything has been installed. 

By the way, are you talking about the main branch or about this particular pull-request branch?

comcon1 commented:
@MagnusSletten I added `runtests.sh` in commit 0412cd5ab6dbbef7f0cf0c0d5f143bcdad21229c

comcon1 commented on Scripts/BuildDatabank/QualityEvaluation.py:
```diff
@@ -54,40 +54,37 @@
-                        lipidExpOPdata = json.load(json_file)
+                    with open(exp_op_fpath) as json_file:
+                        exp_op_data = json.load(json_file)
                 except FileNotFoundError:
```
The idea is to process an exception that I can process, and others will just be forwarded upstairs. But I fixed a bit in b249796d

comcon1 commented on Scripts/DatabankLib/quality.py:
```diff
@@ -252,8 +247,8 @@ def systemQuality(system_fragment_qualities, simulation):
-    for lipid in system_fragment_qualities.keys():
-        _ = get_fragments(simulation.system['COMPOSITION'][lipid]['MAPPING'])
+    for lipid in system_fragment_qualities:
+        _ = get_fragments(simulation.system.content[lipid].mapping_dict)
```
Good point. I've removed the line in b249796d

comcon1 commented on Scripts/DatabankLib/requirements.txt:
```diff
@@ -2,6 +2,9 @@ buildh>=1.6.1
 numpy>=1.20.0,<2.0
 pandas>=2.0.0
-PyYAML>=6.0.0
+yaml
```
Good that you found. Fixed in 0ad6192

MagnusSletten commented:
Ran tests again after the quick requirements fix and everything works great. This is very minor and I suggest we just add it in future pull requests but the output of tests could be cleaned up slightly by registering the markers in the pytest.ini file as described [here](https://docs.pytest.org/en/stable/how-to/mark.html) It just removes some warnings and this is low priority. @comcon1

MagnusSletten reviewed (approved):
Great job! This PR improves the code base by a large margin. Code is clean and well structured with clear intent behind the changes.

batukav reviewed (approved):
My previous comments were mostly on how the exceptions were handled. I see that there's been improvement there. I think the current state is good enough for the merge. We can handle the errors better in the next PR.

comcon1 commented:
I rebased everything on my side removing one conflicting commit regarding the documentation. Then rebase is activated here, and I do it
